### PR TITLE
Zqs 1245/vqa docs

### PR DESCRIPTION
**Pull request by mstechly**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

- Context: why is it needed? In case of bug, what was the cause for the bug?
- Concise description of the implemented solution.
- If any dependencies are added, list them and justify why they are needed.

## Please verify that you have completed the following steps

- [ ] I have self-reviewed my code.
- [ ] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: docs/guides/vqa_guide/introduction.rst [hunk 1]
```diff
@@ -19,7 +19,7 @@ To **estimate** the value of the cost function for given set of parameters, we n
 
 
 TODO: links to ansatzes and estimators.
-:ref:`Optimizers <optimizers_guide>` and :ref:`cost functions <cost_function_guide>` are general features of optimization problems, so this guide focuses on ansatzes and estimators, which are more specific to quantum algorithms. For a more complete, yet approachable discussion, see `Michał's blogpost <https://www.mustythoughts.com/vqas-how-do-they-work>`_.
+:ref:`Optimizers <optimizers_guide>` and :ref:`cost ffunctions <cost_function_guide>` are general features of optimization problems, so this guide focuses on ansatzes and estimators, which are more specific to quantum algorithms. For a more complete, yet approachable discussion, see `Michał's blogpost <https://www.mustythoughts.com/vqas-how-do-they-work>`_.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:22]: typo

File: docs/guides/vqa_guide/introduction.rst [hunk 2]
```diff
@@ -34,9 +34,72 @@ They all work similarly, but each is a little bit different - that's why we will
 
 VQE
 ---
-TODO: check if this is the right header
 
-TODO:
+For most scenarios, the easiest way to use VQE with orquestra is via means of the :class:`orquestra.opt.algorithms.vqe.VQE` class.
+Instances of this class group together several different objects needed for running
+VQEs, and expose convenience methods for finding optimal params and constructing cost functions.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:40]: I'd say `VQE` not `VQEs`

File: docs/guides/vqa_guide/introduction.rst [hunk 3]
```diff
@@ -34,9 +34,72 @@ They all work similarly, but each is a little bit different - that's why we will
...
+The :class:`VQE` class also contains a convenience :meth:`orquestra.opt.algorithms.vqe.VQE.default` method which simplifies creation of its instances even further.
+Since the :meth:`default` method covers most of the use cases, we'll start by describing its arguments:
+
+- :code:`hamilatonian`: a hamiltonian of the problem given in PauliRepresentation.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:44]: Uppercase `Hamiltonian`

Comment (mstechly) [reply docs/guides/vqa_guide/introduction.rst:44]: Api link to `PauliRepresentation` ?

File: docs/guides/vqa_guide/introduction.rst [hunk 4]
```diff
@@ -34,9 +34,72 @@ They all work similarly, but each is a little bit different - that's why we will
...
+
+   opt_value=-0.9999999903697773
+   opt_params=[8.68771769e-01 1.80528665e+00 2.15249204e+00 2.77605314e+00
+    1.57081967e+00 3.14161674e+00 1.28392551e+00 1.66060268e+00
+    2.34053047e+00 1.66618319e+00 1.40691211e-03 1.66622948e+00]
+
+Notice, that you can obtain raw cost function from :class:`VQE` object. This might be useful for instance if you want to compare the obtained result with some other pre-existing solution that you have. Here we verify that the returned optimal value indeed corresponds to the returned params.
+
+.. note::
+   This example only used because we opted for exact computation of expectation values. Obviously,
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:77]: Some words are missing in this sentence?

File: docs/guides/vqa_guide/introduction.rst [hunk 5]
```diff
@@ -34,9 +34,72 @@ They all work similarly, but each is a little bit different - that's why we will
...
+
+   True
+
+As already stated, the default method of constructing :class:`VQE` objects is sufficient for most use cases. Let us now describe how to deal with those cases where the :meth:`default` method is not sufficient.
+
+First of all, there are two ways of customizing the :class:`VQE` objects. You can either construct a new instance using normal initializer of :class:`VQE` object, or start with a default implementation, and then customize it using one of the :meth:`replace_<xyz>` methods. We start with the first method. But before we do, let us explain some simplifications made by the :meth:`default` method.
+
+As you probably guessed, the "uniform" and "proportional" are not the only possible methods of allocating shots. Similarly, there might be other ways of grouping not covered by the :meth:`default` method. Of course, one could think we can extend the method to accept more and more options. However, such code would be rather unpleasant to maintain. And what about your custom methods of allocating shots or grouping? In Orquestra, both of those tasks are implemented as :class:`EstimationPreprocessor` and Under the hood, the :meth:`default` method constructed them for you.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:93]: I'd change to "unpleasant to maintain and confusing to use"?

Comment (mstechly) [reply docs/guides/vqa_guide/introduction.rst:93]: + lower case `Under`

File: docs/guides/vqa_guide/introduction.rst [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+First of all, there are two ways of customizing the :class:`VQE` objects. You can either construct a new instance using normal initializer of :class:`VQE` object, or start with a default implementation, and then customize it using one of the :meth:`replace_xyz` methods. We start with the first method. But before we do, let us explain some simplifications made by the :meth:`default` method.
+
+As you probably guessed, the "uniform" and "proportional" are not the only possible methods of allocating shots. Similarly, there might be other ways of grouping not covered by the :meth:`default` method. Of course, one could think we can extend the :meth:`default` method to accept more and more options. However, such code would be rather unpleasant to maintain and confusing to use. And what about your custom methods of allocating shots or grouping? In Orquestra, both of those tasks are implemented as :class:`orquestra.quantum.api.EstimationPreprocessor`s and under the hood, the :meth:`default` method constructed them for you.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst]: The api link doesn't work here for some reason?

File: docs/guides/vqa_guide/introduction.rst [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+* They store optimizer as one of the attributes. The optimization is done by calling
+  :meth:`find_optimal_params`
+* The instances are supposed to be treated as immutable, but new instances can be constructed by replacing respective attributes using :meth:`replace_xyz` methods.
+
+Knowing those similarities, in the rest of the introduction we will only highlight differences between Orquestra's implementation of VQE and QAOA/QCBM.
+
+QAOA
+----
+
+The Quantum Approximate Optimization Algorithm is implemented by the :class:`QAOA <orquestra.vqa.algorithms.QAOA>` class. The  :meth:`default <orquestra.vqa.algorithms.QAOA.default>` method constructs :class:`QAOA` object using Fahri ansatz with specified number of layers. Using class' initializer, one can further customize what ansatz is used.
```
Comment (mstechly) [inline docs/guides/vqa_guide/introduction.rst:139]: I think it requires more details – e.g. QAOA doesn't have grouping and VQE does.
At least describe what are the inputs for default are and why some are missing from QAOA.

**Comment by max-radin** [comment]
It looks like the Hamiltonian used in the VQE example is non-hermitian (`X0 * Y0 = i Z0`) and so it wouldn't really make sense to use VQE on it. Maybe we should switch to some simple but well-known Hamiltonian? For example transverse Ising model.

**Review by dexter2206 — APPROVED** [review]

**Pull request merged**